Networking Basic Tools Module closes #470 (#471)
bob-crutchley/courseware: topics/networking/modules/basic-tools/README.md

# Basic Networking Tools

<!--TOC_START-->
## Contents
- [Overview](#overview)
	- [`nslookup`](#nslookup)
	- [`ping`](#ping)
- [Tasks](#tasks)
	- [Use `nslookup` to find the IP of a website](#use-nslookup-to-find-the-ip-of-a-website)
	- [Use `ping` to query a site using its IP address](#use-ping-to-query-a-site-using-its-ip-address)
	- [Disconnect your machine during a `ping` request](#disconnect-your-machine-during-a-ping-request)

<!--TOC_END-->
## Overview 

> Two of the most basic tools for seeing networking in action are `nslookup` and `ping`.

Reviewer: ```suggestion
Two of the most basic tools for seeing networking in action are:
- **nslookup**
- **ping**
```

Reviewer: if you use bullet points to list things, you can avoid full-stop at the end of the sentence when highlighting

Both are lookup tools which check the availability of a host.

### `nslookup`

> `nslookup` (name server lookup) is a network administration tool on the command-line, which directly queries the Domain Name System (DNS) to obtain domain name or IP address mapping.

Reviewer: ```suggestion
`nslookup` (**n**ame **s**erver **lookup**) is a network administration tool on the command-line, which directly queries the Domain Name System (DNS) to obtain domain name or IP address mapping.
```

Here, we can check for the **IP address** of a particular Web site using `nslookup`:

```cmd
  C:\Windows\system32>nslookup example.com
  Server:  dns.google
  Address:  8.8.8.8

  Non-authoritative answer:
  Name:    example.com
  Addresses:  2606:2800:220:1:248:1893:25c8:1946
            93.184.216.34
```

> 

Reviewer: ```suggestion

In order to invoke `nslookup`, you need to call it by name (nslookup) in the terminal and pass in a web address (example.com) as an argument to get the information back.

*nslookup* is used in the same way on Windows and Linux operating systems and usually comes pre-installed.

```

This gives us some useful information about `example.com`:

> - The server which `example.com` is hosted on, which is under Google's DNS jurisdiction at `dns.google`.

Reviewer: ```suggestion
- The server which `example.com` is hosted on, which is under Google's DNS jurisdiction at `dns.google`
```

> - Its IPv4 address, which is `93.184.216.34`.

Reviewer: ```suggestion
- Its IPv4 address, which is `93.184.216.34`
```

> - Its IPv6 address, which is `2606:2800:220:1:248:1893:25c8:1946`.

Reviewer: ```suggestion
- Its IPv6 address, which is `2606:2800:220:1:248:1893:25c8:1946`
```

Reviewer: don't end bullet points with a full stop, it's already clear enough where one ends really

The `non-authoritative answer` we see just tells us that `nslookup` is querying DNS records kept on external servers, rather than pinging the entire global DNS system directly.

The two IP addresses refer to the Internet Protocol we mentioned earlier, and just shows us the unique location of a resource (in this case, the `example.com` Web site) across the entire Internet.

More information on IP addresses can be found in the [IP Module](../ip/README.md).
<!--issue-415-->

### `ping`

> We use `ping` to query an IP address so as to check its availability:

Reviewer: expand on what ping is first

A failed `ping` request typically looks like this:
```cmd
  C:\Windows\system32>ping en.wikipedia.org

  Pinging dyna.wikimedia.org [91.198.174.192] with 32 bytes of data:
  Request timed out.
  Request timed out.
  Request timed out.
  Request timed out.

  Ping statistics for 91.198.174.192:
      Packets: Sent = 4, Received = 0, Lost = 4 (100% loss),
```

This is an excellent way to check whether a server is 'up' or 'down', as well as average connection speeds between your machine and the server.

## Tasks

### Use `nslookup` to find the IP of a website
*This task requires the use of `nslookup` inside your machine's command line interface (CLI).*

Use this to find the IP addresses for Amazon, Google, Facebook and Outlook.

### Use `ping` to query a site using its IP address
*This task requires the use of `ping` inside your machine's CLI.*

Use the IP addresses you gained from the previous task to find the DNS which each of the four sites are using. What do you notice? 

### Disconnect your machine during a `ping` request
*This task requires the use of `ping` inside your machine's CLI.*

> Try to `ping` any of the IP addresses from the previous tasks, but either remove your Ethernet cable or deactivate your WiFi during the process. What happens?

Reviewer: ```suggestion
Try to `ping` any of the IP addresses from the previous tasks, but either click `CTRL + C`, remove your Ethernet cable or deactivate your WiFi during the process. What happens?
```
